src/UglyToad.PdfPig/Function/PdfFunctionType0.cs

```diff
@@ -0,0 +1,177 @@
+            : base(functionDictionary, pdfTokenScanner)
+        {
+            if (functionDictionary.TryGet<ArrayToken>(NameToken.Size, pdfTokenScanner, out var size))
+            {
+                Size = size;
+            }
+            else
+            {
+                throw new ArgumentException("Size is Required.");
+            }
+
+            if (functionDictionary.TryGet<NumericToken>(NameToken.BitsPerSample, pdfTokenScanner, out var bitsPerSample))
+            {
+                BitsPerSample = bitsPerSample.Int;
+            }
+            else
+            {
+                throw new ArgumentException("BitsPerSample is Required.");
+            }
+
+            if (functionDictionary.TryGet<NumericToken>(NameToken.Order, pdfTokenScanner, out var order))
+            {
+                Order = order.Int;
+            }
+            else
+            {
+                // set default: 1.
+                Order = 1;
+            }
+
+            if (functionDictionary.TryGet<ArrayToken>(NameToken.Encode, pdfTokenScanner, out var encode))
+            {
+                Encode = encode;
+            }
+            else
+            {
+                // set default: [0 (Size0 − 1) 0 (Size1 − 1) …].
+                List<IToken> values = new List<IToken>();
+                for (int i = 0; i < Size.Length; i++)
+                {
+                    values.Add(new NumericToken(0));
+                    values.Add(new NumericToken(((NumericToken)Size[i]).Int - 1));
+                }
+                Encode = new ArrayToken(values);
+            }
+
+            if (functionDictionary.TryGet<ArrayToken>(NameToken.Decode, pdfTokenScanner, out var decode))
+            {
+                Decode = decode;
+            }
+            else
+            {
+                // set default
+                Decode = Range;
+            }
+        }
+
+        /// <summary>
+        /// PdfFunctionType0
+        /// </summary>
+        /// <param name="functionStream"></param>
+        /// <param name="pdfTokenScanner"></param>
+        public PdfFunctionType0(StreamToken functionStream, IPdfTokenScanner pdfTokenScanner)
+            : this(functionStream.StreamDictionary, pdfTokenScanner)
+        { }
+
+        /// <summary>
+        /// Get the encode for the input parameter.
+        /// </summary>
+        /// <param name="paramNum">The function parameter number.</param>
+        /// <returns>The encode parameter range or null if none is set.</returns>
+        public PdfRange GetEncodeForParameter(int paramNum)
+        {
+            PdfRange retval = null;
+            if (Encode != null && Encode.Length >= paramNum * 2 + 1)
+            {
+                retval = new PdfRange(Encode, paramNum);
+            }
+            return retval;
+        }
+
+        /// <inheritdoc/>
+        public override float[] Eval(float[] input)
+        {
+            throw new NotImplementedException("Eval");
```

Comment:
Do we have PDFBox code we can use to implement these?

Reply:
yes, they are available here: [pdfbox/pdmodel/common/function/](https://github.com/apache/pdfbox/tree/trunk/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/common/function)